pennylane/transforms/optimization/peephole.py

```diff
@@ -0,0 +1,1100 @@
+    sequences are optimized by using 1 qubits and 2 qubits optimal decompositions.
+
+    Args:
+        qfunc (function): A quantum function to be optimized.
+        n_qubits(list(int)): List of size of subset qubits [1], [2], [1, 2], [2, 1]. The order in the list matters as
+                            it will be optimized sequentially.
+        custom_quantum_cost (dict): Optional, custom quantum cost dictionary.
+
+    Returns:
+        function: the transformed quantum function
+
+    Raises:
+        QuantumFunctionError: The subset size list is not of the right format or contains unsupported sizes.
+
+    **Example**
+
+    Consider the following quantum circuit to be optimized
+
+    .. code-block:: python
+
+        def circuit():
+            qml.PauliX(wires=2)
+            qml.CNOT(wires=[1, 2])
+            qml.Hadamard(wires = 1)
+            qml.CNOT(wires=[1, 2])
+            qml.Hadamard(wires = 1)
+            qml.CNOT(wires=[0, 1])
+            qml.CNOT(wires=[0, 2])
+            qml.Hadamard(wires = 2)
+            qml.CNOT(wires=[2, 1])
+            qml.Hadamard(wires = 2)
+            qml.CNOT(wires=[2, 1])
+            qml.PauliX(wires=1)
+            return qml.expval(qml.PauliX(wires=0))
+
+    For optimizing the circuit given the given following template of CNOTs we apply the `pattern_matching`
+    transform.
+
+    >>> dev = qml.device('default.qubit', wires=3)
+    >>> qnode = qml.QNode(circuit, dev)
+    >>> optimized_qfunc = peephole_optimization(n_qubit=[2])(circuit)
+    >>> optimized_qnode = qml.QNode(optimized_qfunc, dev)
+
+    In our case, it is possible to reduce 4 CNOTs to only two CNOTs and therefore
+    optimizing the circuit. The number of CNOTs in the circuit is reduced by one.
+
+    >>> qml.specs(qnode)()["gate_types"]["CNOT"]
+    6
+
+    >>> qml.specs(optimized_qnode)()["gate_types"]["CNOT"]
+    4
+
+    >>> print(qml.draw(qnode)())
+    0: ────────────────╭C─╭C────────────────┤  <X>
+    1: ────╭C──H─╭C──H─╰X─│─────╭X────╭X──X─┤
+    2: ──X─╰X────╰X───────╰X──H─╰C──H─╰C────┤
+
+    >>> print(qml.draw(optimized_qnode)())
+    0: ─╭C───────────────────────────────────────────────────────╭C─┤  <X>
+    1: ─│───RZ(1.57)─────────────╭C──SX─╭C──Rot(-1.57,1.57,1.57)─╰X─┤
+    2: ─╰X──Rot(0.00,1.57,-1.57)─╰X──S──╰X──Rot(0.00,1.57,-1.57)────┤
+
+    The algortihm finds the sequence [0, 1, 2, 3, 4, 7, 8, 9, 10, 11] (acting on qubits 1 and 2) as we can push CNOT
+    5 to the right and CNOT 6 to the left. The longest sequence contains 4 CNOT which is not optimal and can bbe
+    reduced to two by optimal synthesis of 4x4 unitaries. Therefore the sequences of gates is replaced and therefore
+    the whole circuit is optimized.
+
+    **Reference:**
+
+    [1] Iten, R., Moyard, R., Metger, T., Sutter, D. and Woerner, S., 2022.
+    Exact and practical pattern matching for quantum circuit optimization.
+    `doi.org/10.1145/3498325 <https://dl.acm.org/doi/abs/10.1145/3498325>`_
```

Comment:
I don't know if it's included in this reference, but if not it might be nice include references for the optimal synthesis algorithm of 1 and 2 qubit unitaries.